File: docs/eos/examples/compliance-starter/compliance_test.go
// compliance_test.go — proves the Go skeletons run out of the box. Zero-dependency (stdlib
// `testing`). Run: `go test ./...` (or `go test -v`) from this directory.
// Mirrors compliance.test.mjs / test_compliance.py one-for-one (same 7 cases).
package compliance

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
)

func TestRedactionRegimeScoping(t *testing.T) {
	// PCI-DSS scope = cardholder data only; a plain email is out of scope and stays.
	pci := CreateRedactor([]string{"PCI-DSS"}) // alias resolves to PCI
	p := pci.Redact(map[string]any{"cardNumber": "4111111111111111", "email": "jane@example.com"}).(map[string]any)
	if p["cardNumber"] != "«REDACTED:cardnumber»" {
		t.Fatalf("PCI should mask cardNumber, got %v", p["cardNumber"])
	}
	if p["email"] != "jane@example.com" {
		t.Fatalf("PCI scope must not touch a non-card email, got %v", p["email"])
	}

	// HIPAA scope = PHI (identifiers + health), but NOT a raw card number (that is PCI's regime).
	hipaa := CreateRedactor([]string{"HIPAA"})
	h := hipaa.Redact(map[string]any{"mrn": "MR-9", "email": "jane@example.com", "pan": "4111111111111111"}).(map[string]any)
	if h["mrn"] != "«REDACTED:mrn»" {
		t.Fatalf("HIPAA should mask mrn, got %v", h["mrn"])
	}
	if h["email"] != "«REDACTED:email»" {
		t.Fatalf("HIPAA should mask email, got %v", h["email"])
	}
	if h["pan"] != "4111111111111111" {
		t.Fatalf("HIPAA-only must not redact card data (select PCI for that), got %v", h["pan"])
	}

	// base-only (no regime) = credentials/secrets only.
	base := CreateRedactor([]string{})
	if got := base.Redact(map[string]any{"token": "abc123", "email": "a@b.co"}).(map[string]any)["token"]; got != "«REDACTED:token»" {
		t.Fatalf("base should mask token, got %v", got)
	}
	if got := base.Redact(map[string]any{"email": "a@b.co"}).(map[string]any)["email"]; got != "a@b.co" {
		t.Fatalf("base must not touch email, got %v", got)
	}

	if len(pci.Regimes) != 1 || pci.Regimes[0] != "PCI" {
		t.Fatalf("expected regimes [PCI], got %v", pci.Regimes)
	}
	for _, n := range []string{"HIPAA", "PCI", "GDPR_PIPL"} {
		if _, ok := Profiles[n]; !ok {
			t.Fatalf("named regime preset %q must be exported", n)
		}
	}
}

func TestRedactionMasksCore(t *testing.T) {
	out := Redact(map[string]any{
		"userId": "u-123", // safe identifier — kept
		"email":  "jane@example.com",
		"note":   "card 4111 1111 1111 1111, ssn 123-45-6789, call 555-123-4567",
		"order":  map[string]any{"total": 42}, // safe — kept
	}).(map[string]any)
	if out["userId"] != "u-123" {
		t.Fatalf("safe id changed: %v", out["userId"])
	}
	if out["order"].(map[string]any)["total"] != 42 {
		t.Fatalf("nested safe value changed: %v", out["order"])
	}
	if out["email"] != "«REDACTED:email»" {
		t.Fatalf("email not masked: %v", out["email"])
	}
	note := out["note"].(string)
	if strings.Contains(note, "4111") {
		t.Fatal("PAN not stripped")
	}
	if strings.Contains(note, "123-45-6789") {
		t.Fatal("SSN not stripped")
	}
	if strings.Contains(note, "555-123-4567") {
		t.Fatal("phone not stripped")
	}
}

func TestLuhn(t *testing.T) {
	if !LuhnValid("4111111111111111") {
		t.Fatal("valid PAN should pass Luhn")
	}
	if LuhnValid("1234567890123456") {
		t.Fatal("non-card 16-digit id should fail Luhn")
	}
}

func TestAssertCleanBoundaryGuard(t *testing.T) {
	if AssertClean(map[string]any{"ssn": "123-45-6789"}) == nil {
		t.Fatal("assertClean should flag an ssn leak")
	}
	if len(Scan(map[string]any{"userId": "u-1", "qty": 3})) != 0 {
		t.Fatal("clean payload should scan empty")
	}
	if AssertClean(Redact(map[string]any{"email": "a@b.co", "qty": 3})) != nil {
		t.Fatal("payload should be clean after redact")
	}
}

func TestConsent(t *testing.T) {
	c := NewConsentStore(nil)
	c.Grant("u-1", "marketing")
	c.Grant("u-1", "analytics")
	if !c.Check("u-1", "marketing") {
		t.Fatal("marketing should be granted")
	}

	c.Revoke("u-1", "marketing") // withdrawal as easy as granting
	if c.Check("u-1", "marketing") {
		t.Fatal("marketing should be revoked")
	}
	if !c.Check("u-1", "analytics") {
		t.Fatal("purposes must be independent (PIPL separate consent)")
	}
	if c.Check("u-1", "never-asked") {
		t.Fatal("never-asked purpose must be false")
	}

	keys := []string{}
	for k := range c.State("u-1") {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	if len(keys) != 2 || keys[0] != "analytics" || keys[1] != "marketing" {
		t.Fatalf("expected [analytics marketing], got %v", keys)
	}
	if n := len(c.Log.All()); n != 3 {
		t.Fatalf("expected 3 audit entries (2 grants + 1 revoke), got %d", n)
	}
}

// fakeSource is a DSAR adapter for the test: name + export + erase, recording erase order.
type fakeSource struct {
	name  string
	rows  any
	count int
	sink  *[]string
}

func (f fakeSource) Name() string      { return f.name }
func (f fakeSource) Export(string) any { return f.rows }
func (f fakeSource) Erase(string) int  { *f.sink = append(*f.sink, f.name); return f.count }

func TestDSAR(t *testing.T) {
	var erased []string
	sources := []Source{
		fakeSource{name: "profiles", rows: []any{map[string]any{"id": "u-1", "city": "X"}}, count: 1, sink: &erased},
		fakeSource{name: "orders", rows: []any{map[string]any{"id": "o-9"}}, count: 1, sink: &erased},
	}

	bundle := ExportSubject("u-1", sources, nil)
	if _, ok := bundle.Data["profiles"]; !ok {
		t.Fatal("export must include profiles source")
	}
	if _, ok := bundle.Data["orders"]; !ok {
		t.Fatal("export must include orders source")
	}
	if bundle.GeneratedAt == "" {
		t.Fatal("export must be timestamped/portable")
	}

	receipt := EraseSubject("u-1", sources, "user-request", nil)
	sort.Strings(erased)
	if len(erased) != 2 || erased[0] != "orders" || erased[1] != "profiles" {
		t.Fatalf("erase must run each source, got %v", erased)
	}
	if receipt.Erased["profiles"] != 1 || receipt.Erased["orders"] != 1 {
		t.Fatalf("erase counts wrong: %v", receipt.Erased)
	}
	if receipt.Reason != "user-request" {
		t.Fatalf("reason not recorded: %v", receipt.Reason)
	}
}

func TestProfileAutoSelect(t *testing.T) {
	// Pure parse: canonical line -> resolved profile names; rationale words are ignored.
	if got := strings.Join(ParseRegimes("# Compliance profile\n\n**Regulatory regime:** HIPAA, PCI-DSS\n"), ","); got != "HIPAA,PCI" {
		t.Fatalf("multi-regime parse wrong: %q", got)
	}
	// Slash / plus separators and mixed case also resolve.
	if got := strings.Join(ParseRegimes("Regulatory regime: gdpr / pipl"), ","); got != "GDPR_PIPL" {
		t.Fatalf("gdpr/pipl parse wrong: %q", got)
	}
	// Explicit `none` (with parenthetical rationale naming a regime) -> base only (empty).
	if got := ParseRegimes("Regulatory regime: none (generic PII, no PCI applies)"); len(got) != 0 {
		t.Fatalf("none must parse to empty, got %v", got)
	}

	// File round-trip: write a profile, build the redactor straight from it.
	dir := t.TempDir()
	f := filepath.Join(dir, "compliance-profile.md")
	if err := os.WriteFile(f, []byte("# Compliance profile\n\n**Regulatory regime:** HIPAA\n\nRationale: PHI.\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	r, err := RedactorFromProfile(f)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if strings.Join(r.Regimes, ",") != "HIPAA" {
		t.Fatalf("redactor should scope to HIPAA, got %v", r.Regimes)
	}

	// Missing file -> fail-safe: all regimes + the read error (never silently under-redact).
	strict, err := RedactorFromProfile(filepath.Join(dir, "nope.md"))
	if err == nil {
		t.Fatal("missing profile should return the read error")
	}
	if strings.Join(strict.Regimes, ",") != "HIPAA,PCI,GDPR_PIPL" {
		t.Fatalf("missing profile must fail safe to all regimes, got %v", strict.Regimes)
	}
}

File: docs/eos/examples/compliance-starter/ComplianceTest.java
// ComplianceTest.java — proves the Java skeletons run out of the box. Zero-dependency: a plain
// main() harness (no JUnit, no build tool), mirroring how the Node/Python ports use their
// built-in runners. Mirrors compliance.test.mjs / test_compliance.py one-for-one (same 7 cases).
//
// Compile + run (UTF-8 for the «REDACTED» mask):
//   javac -encoding UTF-8 *.java && java ComplianceTest
//
// In a real project you'd promote these to JUnit 5; the shapes are identical.
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class ComplianceTest {
    private static int passed = 0;
    private static int failed = 0;

    public static void main(String[] args) {
        redactionRegimeScoping();
        redactionMasksCore();
        luhn();
        assertCleanBoundaryGuard();
        consent();
        dsar();
        profileAutoSelect();

        System.out.println();
        System.out.println((failed == 0 ? "OK" : "FAILED")
                + " — " + passed + " checks passed, " + failed + " failed across 7 tests");
        if (failed > 0) {
            System.exit(1);
        }
    }

    private static void check(boolean cond, String msg) {
        if (cond) {
            passed++;
        } else {
            failed++;
            System.out.println("  \u2717 FAIL: " + msg);
        }
    }

    /** Build an ordered String->Object map from k, v, k, v, ... pairs. */
    private static Map<String, Object> map(Object... kv) {
        Map<String, Object> m = new LinkedHashMap<>();
        for (int i = 0; i + 1 < kv.length; i += 2) {
            m.put((String) kv[i], kv[i + 1]);
        }
        return m;
    }

    @SuppressWarnings("unchecked")
    private static void redactionRegimeScoping() {
        // PCI-DSS scope = cardholder data only; a plain email is out of scope and stays.
        Redaction.Redactor pci = Redaction.createRedactor(List.of("PCI-DSS")); // alias resolves to PCI
        Map<String, Object> p = (Map<String, Object>) pci.redact(
                map("cardNumber", "4111111111111111", "email", "jane@example.com"));
        check("«REDACTED:cardnumber»".equals(p.get("cardNumber")), "PCI masks card number");
        check("jane@example.com".equals(p.get("email")), "PCI scope must not touch a non-card email");

        // HIPAA scope = PHI (identifiers + health), but NOT a raw card number (that is PCI's regime).
        Redaction.Redactor hipaa = Redaction.createRedactor(List.of("HIPAA"));
        Map<String, Object> h = (Map<String, Object>) hipaa.redact(
                map("mrn", "MR-9", "email", "jane@example.com", "pan", "4111111111111111"));
        check("«REDACTED:mrn»".equals(h.get("mrn")), "HIPAA masks mrn");
        check("«REDACTED:email»".equals(h.get("email")), "HIPAA masks email");
        check("4111111111111111".equals(h.get("pan")), "HIPAA-only must not redact card data (select PCI)");

        // base-only (no regime) = credentials/secrets only.
        Redaction.Redactor base = Redaction.createRedactor(List.of());
        Map<String, Object> b1 = (Map<String, Object>) base.redact(map("token", "abc123", "email", "a@b.co"));
        check("«REDACTED:token»".equals(b1.get("token")), "base masks a credential token");
        Map<String, Object> b2 = (Map<String, Object>) base.redact(map("email", "a@b.co"));
        check("a@b.co".equals(b2.get("email")), "base must not touch a non-credential email");

        check(pci.regimes.equals(List.of("PCI")), "reports the active regime set");
        check(Redaction.PROFILES.containsKey("HIPAA") && Redaction.PROFILES.containsKey("PCI")
                && Redaction.PROFILES.containsKey("GDPR_PIPL"), "named regime presets are exported");
    }

    @SuppressWarnings("unchecked")
    private static void redactionMasksCore() {
        Map<String, Object> out = (Map<String, Object>) Redaction.redact(map(
                "userId", "u-123", // safe identifier — kept
                "email", "jane@example.com",
                "note", "card 4111 1111 1111 1111, ssn 123-45-6789, call 555-123-4567",
                "order", map("total", 42))); // safe — kept
        check("u-123".equals(out.get("userId")), "safe identifier kept");
        check(Integer.valueOf(42).equals(((Map<String, Object>) out.get("order")).get("total")),
                "nested safe value kept");
        check("«REDACTED:email»".equals(out.get("email")), "email masked");
        String note = (String) out.get("note");
        check(!note.contains("4111"), "PAN stripped");
        check(!note.contains("123-45-6789"), "SSN stripped");
        check(!note.contains("555-123-4567"), "phone stripped");
    }

    private static void luhn() {
        check(Redaction.luhnValid("4111111111111111"), "valid PAN passes Luhn");
        check(!Redaction.luhnValid("1234567890123456"), "non-card 16-digit id fails Luhn");
    }

    private static void assertCleanBoundaryGuard() {
        boolean threw = false;
        try {
            Redaction.assertClean(map("ssn", "123-45-6789"));
        } catch (RuntimeException e) {
            threw = true;
        }
        check(threw, "assertClean throws on a leak");
        check(Redaction.scan(map("userId", "u-1", "qty", 3)).isEmpty(), "clean payload scans empty");
        boolean threwAfterRedact = false;
        try {
            Redaction.assertClean(Redaction.redact(map("email", "a@b.co", "qty", 3)));
        } catch (RuntimeException e) {
            threwAfterRedact = true;
        }
        check(!threwAfterRedact, "assertClean passes on redacted payload");
    }

    private static void consent() {
        Consent c = Consent.create();
        c.grant("u-1", "marketing");
        c.grant("u-1", "analytics");
        check(c.check("u-1", "marketing"), "marketing granted");

        c.revoke("u-1", "marketing"); // withdrawal as easy as granting
        check(!c.check("u-1", "marketing"), "marketing revoked");
        check(c.check("u-1", "analytics"), "purposes independent (PIPL separate consent)");
        check(!c.check("u-1", "never-asked"), "never-asked purpose is false");

        List<String> keys = new ArrayList<>(c.state("u-1").keySet());
        Collections.sort(keys);
        check(keys.equals(List.of("analytics", "marketing")), "state lists both purposes");
        check(c.log.all().size() == 3, "audit trail has 3 events (2 grants + 1 revoke), no PII");
    }

    private static void dsar() {
        List<String> erased = new ArrayList<>();
        List<Dsar.Source> sources = List.of(
                source("profiles", List.of(map("id", "u-1", "city", "X")), 1, erased),
                source("orders", List.of(map("id", "o-9")), 1, erased));

        Map<String, Object> bundle = Dsar.exportSubject("u-1", sources, null);
        @SuppressWarnings("unchecked")
        Map<String, Object> data = (Map<String, Object>) bundle.get("data");
        check(data.containsKey("profiles") && data.containsKey("orders"), "export gathers all sources");
        check(bundle.get("generatedAt") != null, "export is timestamped/portable");

        Map<String, Object> receipt = Dsar.eraseSubject("u-1", sources, "user-request", null);
        List<String> sorted = new ArrayList<>(erased);
        Collections.sort(sorted);
        check(sorted.equals(List.of("orders", "profiles")), "erase runs each source");
        @SuppressWarnings("unchecked")
        Map<String, Object> er = (Map<String, Object>) receipt.get("erased");
        check(Integer.valueOf(1).equals(er.get("profiles")) && Integer.valueOf(1).equals(er.get("orders")),
                "erase counts recorded");
        check("user-request".equals(receipt.get("reason")), "reason recorded");
    }

    private static void profileAutoSelect() {
        // Pure parse: canonical line -> resolved profile names; rationale words are ignored.
        check(Redaction.parseRegimes("# Compliance profile\n\n**Regulatory regime:** HIPAA, PCI-DSS\n")
                .equals(List.of("HIPAA", "PCI")), "parse multi-regime line");
        // Slash / plus separators and mixed case also resolve.
        check(Redaction.parseRegimes("Regulatory regime: gdpr / pipl").equals(List.of("GDPR_PIPL")),
                "parse gdpr/pipl");
        // Explicit `none` (with parenthetical rationale naming a regime) -> base only (empty).
        check(Redaction.parseRegimes("Regulatory regime: none (generic PII, no PCI applies)").isEmpty(),
                "none parses to empty");

        try {
            // File round-trip: write a profile, build the redactor straight from it.
            Path f = Files.createTempFile("eos-profile-", ".md");
            try {
                Files.writeString(f, "# Compliance profile\n\n**Regulatory regime:** HIPAA\n\nRationale: PHI.\n");
                check(Redaction.redactorFromProfile(f.toString()).regimes.equals(List.of("HIPAA")),
                        "redactorFromProfile scopes to the profile regime");
            } finally {
                Files.deleteIfExists(f);
            }

            // Missing file -> fail-safe: all regimes (never silently under-redact).
            String missing = f.getParent().resolve("eos-nope-" + System.nanoTime() + ".md").toString();
            check(Redaction.redactorFromProfile(missing).regimes.equals(List.of("HIPAA", "PCI", "GDPR_PIPL")),
                    "missing profile fails safe to all regimes");
            // Opt-in softer fallbacks.
            check(Redaction.redactorFromProfile(missing, "base").regimes.isEmpty(), "fallback=base -> base only");
            boolean threw = false;
            try {
                Redaction.redactorFromProfile(missing, "throw");
            } catch (RuntimeException e) {
                threw = true;
            }
            check(threw, "fallback=throw raises");
        } catch (IOException e) {
            check(false, "temp-file round-trip: " + e.getMessage());
        }
    }

    /** A DSAR source adapter: name + export(subject) + erase(subject), recording erase order. */
    private static Dsar.Source source(String name, Object rows, int count, List<String> sink) {
        return new Dsar.Source() {
            @Override
            public String name() {
                return name;
            }

            @Override
            public Object export(String subjectId) {
                return rows;
            }

            @Override
            public int erase(String subjectId) {
                sink.add(name);
                return count;
            }
        };
    }

    private ComplianceTest() {
    }
}
